OZprivate/rawJS/OZTreeModule/src/render/renderer.js

```diff
@@ -110,6 +134,11 @@ function need_refresh() {
 
   if (render_id % 60 === 0) return true;
   if (tree_state.xp != last_xp || tree_state.yp != last_yp || tree_state.ws != last_ws) return true;
+  if (tree_state.button_x != last_button_x || tree_state.button_y != last_button_y) {
+    last_button_x = tree_state.button_x;
+    last_button_y = tree_state.button_y;
+    return true;
```

Comment:
In principle I expected that we'd be able to call get_shapes in here rather than returning true, since that's (as far as I understand) the thing that is actually updating global button action and then the next line would return true if necessary, which would be obviously better, but I couldn't get that to work at first glance, it just got stuck not rendering at all.

Reply:
I'd assume there's something seeing the second call to ``get_shapes()`` with an identical tree and deciding rendering doesn't need to happen, but what that is I dunno.

Anyway, how about not bothering with ``last_button_x``, and moving the ``get_shapes()`` call outside of ``if (need_refresh())``?

https://github.com/OneZoom/OZtree/blob/5eff75258c3962c22cb9396ad5f2e64c1635c821/OZprivate/rawJS/OZTreeModule/src/render/renderer.js#L59-L60

This is more wasteful admittedly, as we're then running ``get_shapes()`` on every refresh rather than every refresh with a moving mouse cursor. But it's just ``get_shapes()`` not ``refresh_by_redraw()``, and if there's no activity the whole refresh loop will fall idle & stop anyway.

We could make it 2 stage maybe, a ``need_shapes()`` & ``need_refresh()``, but a quick profile shows that ``get_shapes()`` suggests that ``get_shapes`` isn't that expensive in comparison to the actual render (which is mostly slowed down by text measuring).

Reply:
Hmm. Trying that, I'm finding I get a fiar bit of garbage collection cruft and stuttering when running at 120Hz. I'm not being very scientific about it though.
I've pushed a commit to pull the get_shapes outside of the need_refresh only when the cursor has moved which seems to work okay. See what you think of that.